Categorical type

TomAugspurger opened the pull request:
We extended the CategoricalDtype to accept optional `categories` and `ordered`
argument. CategoricalDtype is now part of the public API. This allows users to
specify the desired categories and orderedness of an operation ahead of time.
The current behavior, which is still possible with `categories=None`, the
default, is to infer the categories from whatever is present.

This change will make it easy to implement support for specifying categories
that are know ahead of time in other places e.g. `.astype`, `.read_csv`, and the
`Series` constructor.

Closes #14711 
Closes #15078
Closes #14676 

Dunno if we would want to do this for 0.20 or not. It ended up not being *too* big a change, so maybe.

jreback commented:
.dtype on a Categorical/CI should construct this

Categorica.is_dtype_equal should just work

jreback commented:
can u run asv on this to make sure nothing much changing perf wise

TomAugspurger commented:
> .dtype on a Categorical/CI should construct this

Ha, that's why it felt like so much less code this time, because I forgot to do the other half :)

I'll work on this a bit more today, but will probably push to 0.21

jreback commented:
yeah this feels like it should be in 0.21.0. *maybe* we also expose this in ``pandas.api.types.CategoricalDtype`` rather than at the top-level.

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -131,6 +171,33 @@ def __eq__(self, other):
 
+    @staticmethod
+    def _hash_categories(categories):
+        from pandas.tools.hashing import hash_array, _combine_hash_arrays
```
you can just use``hash_pandas_object``. Its an ordered hash of the data.

gfyoung commented on pandas/tests/core/dtypes/test_dtypes.py:
```diff
@@ -356,6 +357,70 @@ def test_not_string(self):
         self.assertFalse(is_string_dtype(PeriodDtype('D')))
 
 
+class TestCategoricalDtypeParametrized(object):
```
I think we can just call this `TestCategoricalDtype` (consistent with other class names)

gfyoung commented on pandas/tests/series/test_constructors.py:
```diff
@@ -148,6 +149,30 @@ def test_constructor_categorical(self):
+        result = pd.Series(['a', 'b'], dtype=pd.CategoricalDtype(['b', 'a']))
+        assert is_categorical_dtype(result)
+        tm.assert_index_equal(result.cat.categories, pd.Index(['b', 'a']))
+        assert result.cat.ordered is False
```
What's more idiomatic: `assert ... is False` or `assert not ...` ?

jreback replied:
if you are asserting that the result is a boolean (``False``) , then the first is more appropriate. If its just not ``True`` then the 2nd is fine (e.g. imagine 0 satisfies 2nd but not 1st). In this case we *do* want to assert a boolean result.

gfyoung commented on pandas/tests/core/dtypes/test_dtypes.py:
```diff
@@ -356,6 +357,70 @@ def test_not_string(self):
+    def test_categories(self):
+        result = CategoricalDtype(['a', 'b', 'c'])
+        tm.assert_index_equal(result.categories, pd.Index(['a', 'b', 'c']))
+        assert result.ordered is False
```
Same comment as <a href="https://github.com/pandas-dev/pandas/pull/16015/files#r111701915>here</a>

pep8speaks commented:
Hello @TomAugspurger! Thanks for updating the PR.



 - In the file [`pandas/core/categorical.py`](https://github.com/pandas-dev/pandas/blob/04a77d38c0dc6e08e08df191fd5b27e0cd52cb86/pandas/core/categorical.py), following are the PEP8 issues :

> [Line 334:17](https://github.com/pandas-dev/pandas/blob/04a77d38c0dc6e08e08df191fd5b27e0cd52cb86/pandas/core/categorical.py#L334): [E129](https://duckduckgo.com/?q=pep8%20E129) visually indented line with same indent as next logical line

TomAugspurger commented on pandas/tests/dtypes/test_dtypes.py:
```diff
@@ -618,3 +618,8 @@ def test_categorical_equality(self, ordered, other, expected):
         result = c1 == other
         assert result == expected
+
+    def test_mixed(self):
```
This is going to fail at the moment. `hash_object_array` tries to work on the mixed array first, and if that fails it falls back to `hash_object_array(a.astype(str))`, which will cause a "collision" with the section set of categoricals. Not sure how I'll handle this yet.

jreback commented on doc/source/categorical.rst:
```diff
@@ -140,6 +142,54 @@ constructor to save the factorize step during normal constructor mode:
+CategoricalDtype
+----------------
+
+.. versionadded:: 0.21.0
```
versionchanged; it exists now.

jreback commented on pandas/core/api.py:
```diff
@@ -6,6 +6,7 @@
 
 from pandas.core.algorithms import factorize, unique, value_counts
 from pandas.core.dtypes.missing import isna, isnull, notna, notnull
```
I don't think exposing this is a good idea here. We don't directly do this with other dtypes. instead the are in ``pandas.api.types``

TomAugspurger replied:
Hmm, I could go either way on this. It would be the only type at the top level. On the other hand `pd.api.types.CategoricalDtype(values)` is quite long :)

I'll think about it some more, but you're probably right.

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -1671,7 +1671,9 @@ def _coerce_to_dtype(dtype):
-        dtype = CategoricalDtype()
+        categories = getattr(dtype, 'categories', None)
+        ordered = getattr(dtype, 'ordered', False)
+        dtype = CategoricalDtype(categories=categories, ordered=ordered)
```
this is not really necessary if you allow ``CategoricalDtype(dtype)`` IOW, you can check whether the incoming ``categories`` is an instance of a ``CagetegoricalDtype`` and act accordingly.

jreback replied:
?

TomAugspurger replied:
I'm not sure I like overloading the `categories` like that.

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -110,37 +111,147 @@ class CategoricalDtypeType(type):
+    Type for categorical data with the categories and orderedness,
+    but not the values.
+
+    .. versionadded:: 0.20.0
```
versionchanged

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -110,37 +111,147 @@ class CategoricalDtypeType(type):
+    Examples
+    --------
+    >>> t = CategoricalDtype(categories=['b', 'a'], ordered=True)
+    >>> s = Series(['a', 'a', 'b', 'b', 'a'])
```
I wouldn't show anything not related to the actual dtype here.

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -110,37 +111,147 @@ class CategoricalDtypeType(type):
+
+    def __new__(cls, categories=None, ordered=False, fastpath=False):
+        from pandas.core.indexes.base import Index
+        if categories is not None:
```
do the import inside the function

jreback replied:
categories *could* be a CategoricalDtype (see my comment above)

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -110,37 +111,147 @@ class CategoricalDtypeType(type):
+            # validation
+            cls._validate_categories(categories, fastpath=fastpath)
+            cls._validate_ordered(ordered)
+            # We have a choice when hashing pandas unordered categoricals
```
use blank lines to indicate different thoughts

jreback reviewed (changes requested):
looks pretty good. I don't like the MultiIndex change (already commened on that PR). instead I would simply just change it (rather than adding a kw); the small api break is fine.

jorisvandenbossche commented:
@TomAugspurger just to check: this is ready for review?

TomAugspurger commented:
It's close. This weekend probably. I'll submit a comment with a summary of the changes and let you know when it's really ready.

________________________________
From: Joris Van den Bossche <notifications@github.com>
Sent: Friday, August 18, 2017 5:26:43 PM
To: pandas-dev/pandas
Cc: Tom Augspurger; Mention
Subject: Re: [pandas-dev/pandas] Categorical type (#16015)


@TomAugspurger<https://github.com/tomaugspurger> just to check: this is ready for review?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pandas-dev/pandas/pull/16015#issuecomment-323474370>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABQHIopwBcbW-elgQJliYxJ-ItR4Idswks5sZg-jgaJpZM4M-hoW>.

TomAugspurger committed:
ENH: Parametrized CategoricalDtype

We extended the CategoricalDtype to accept optional categories and ordered
argument.

```python
pd.CategoricalDtype(categories=['a', 'b'], ordered=True
```

CategoricalDtype is now part of the public API. This allows users to
specify the desired categories and orderedness of an operation ahead of time.
The current behavior, which is still possible with categories=None, the
default, is to infer the categories from whatever is present.

This change will make it easy to implement support for specifying categories
that are know ahead of time in other places e.g. .astype, .read_csv, and the
Series constructor.

Closes #14711
Closes #15078
Closes #14676

TomAugspurger commented:
Alighty, this should be more or less ready for review. I have the PR split into two parts (well 3 if you count the cherry-picked 1abaecb from a separate PR)

a7eb835 has all the "new" stuff, along with tests for the new things.

f9236b4 has all the "API bending" stuff, the tests I had to update.

jreback commented on doc/source/advanced.rst:
```diff
@@ -654,7 +654,7 @@ setting the index of a ``DataFrame/Series`` with a ``category`` dtype would conv
    df = pd.DataFrame({'A': np.arange(6),
                       'B': list('aabbca')})
-   df['B'] = df['B'].astype('category', categories=list('cab'))
+   df['B'] = df['B'].astype(pd.CategoricalDtype(list('cab')))
```
I think i left a comment about this. but this ``CategoricalDtype`` is already in ``pandas.api.types``. do not add to the top level.

TomAugspurger replied:
Fixed in 1364f2d

jreback commented on doc/source/categorical.rst:
```diff
@@ -140,6 +147,61 @@ constructor to save the factorize step during normal constructor mode:
     splitter = np.random.choice([0,1], 5, p=[0.5,0.5])
     s = pd.Series(pd.Categorical.from_codes(splitter, categories=["train", "test"]))
 
```
add a ref

jreback reviewed (changes requested):
i need to fully review.

TomAugspurger commented:
@jreback I'm looking into the performance now. Just micro-benchmarks at the moment, I'll follow up with more realistic ones, but on these two:

```python
# 1. single:
CategoricalDtype(range(100))

# 2. cached:
data = range(100)
x = CategoricalDtype(data)  # in the cache
CategoricalDtype(data)
```

I'm comparing two versions: caching (this branch) and no caching (essentially the same, but without the hashing of values / caching in `CategoricalDtype.__new__`).

| Kind   | Caching | No-Caching |
| ------ | ------- | ---------- |
| single | 125     | 35.6       |
| cached | 116     | 35         |

Hashing takes up about 60% of the time in the `CategoricalDtype.__new__`, so it'd be nice to avoid if possible.

---

What kind of restrictions do we have on instances of extension types being singletons (per parameter). I've been basing this off `PeriodDtype`, and we test for it, but how important is it?

TomAugspurger commented:
I've gone ahead and removed the caching of `{hash: CategoricalDtype}` in https://github.com/pandas-dev/pandas/pull/16015/commits/e8dda56552a2c38b7d80c9b3a5188de815e60008

I like that it's simpler and faster, but can certainly add that back in if required. Clearly there were some issues with the hashing / caching implementation in the [old CI runs](https://travis-ci.org/pandas-dev/pandas/jobs/270193477#L1183)

TomAugspurger commented:
OK, the only windows failure was on fastparquet [here](https://ci.appveyor.com/project/pandas-dev/pandas/build/1.0.4133/job/7heo6k1pjs3a6up3#L1670) has been fixed and will be in the next release of fastparqeut (next week I think). I can either skip for now or just wait till that's out.

TomAugspurger commented:
@jreback @jorisvandenbossche gentle ping if you don't mind giving this a look. It'd be nice to have in master for a little while before the release.

I've skipped the fastparquet categorical tests for versions < 0.1.3 (which is all versions but master at the moment). I've tested locally against fastparquet master.

jreback commented:
will have a look tomorrow

jreback commented on doc/source/advanced.rst:
```diff
@@ -649,7 +649,7 @@ setting the index of a ``DataFrame/Series`` with a ``category`` dtype would conv
    df = pd.DataFrame({'A': np.arange(6),
                       'B': list('aabbca')})
-   df['B'] = df['B'].astype('category', categories=list('cab'))
+   df['B'] = df['B'].astype(pd.api.types.CategoricalDtype(list('cab')))
```
I would leave this here

TomAugspurger replied:
I'm not a huge fan of the overloading of `.astype` signature, so I removed all instances where we pass through any keyword args. We should still have an example of just `.astype('category')`, with no kwargs. If you want I can change this back though.

jreback replied:
ok that's fine, it just reads better if you import ``CategoricalDtype``, e.g.

``df['B'].astype(CategoricalDtype(list('cab')))``

jreback replied:
I suppose we could deprecate passing the ``categories`` thru ``.astype`` as well

jreback commented on doc/source/categorical.rst:
```diff
@@ -140,6 +148,62 @@ constructor to save the factorize step during normal constructor mode:
     splitter = np.random.choice([0,1], 5, p=[0.5,0.5])
     s = pd.Series(pd.Categorical.from_codes(splitter, categories=["train", "test"]))
 
+CategoricalDtype
```
add a sub-section ref here

jreback commented on doc/source/categorical.rst:
```diff
@@ -140,6 +148,62 @@ constructor to save the factorize step during normal constructor mode:
+----------------
+
+.. versionchanged:: 0.21.0
+
```
maybe add these as bullet points

jreback commented on doc/source/categorical.rst:
```diff
@@ -140,6 +148,62 @@ constructor to save the factorize step during normal constructor mode:
+.. ipython:: python
+
+   c1 = pd.api.types.CategoricalDtype(['a', 'b', 'c'], ordered=False)
+   # Equal, since order is not considered when ordered=False
```
add a blank line before comments

jreback commented on doc/source/categorical.rst:
```diff
@@ -140,6 +148,62 @@ constructor to save the factorize step during normal constructor mode:
+   c1 == 'category'
+
+
+.. warning::
```
this is confusing, better as a comment in the code itself.

TomAugspurger replied:
You're saying not in the user-docs at all? I think it's worthwhile including precisely becuas it's so confusing (I also have a comment in `CategoricalDtype.__eq__` explaining this.

jreback replied:
I think its a very subtle point and will be lost on the user. If you can show a case where the user might be confused then the warning would be ok.

jreback commented on doc/source/categorical.rst:
```diff
@@ -189,7 +253,9 @@ It's also possible to pass in the categories in a specific order:
 
-         s = pd.Series(list('babc')).astype('category', categories=list('abcd'))
+         s = pd.Series(list('babc')).astype(
+             pd.api.types.CategoricalDtype(list('abcd'))
```
I would import ``CategoricalDtype`` at the top of ``categorical.rst`` (or in an appropriate place) I think.

jreback commented on doc/source/merging.rst:
```diff
@@ -831,7 +831,7 @@ The left frame.
 .. ipython:: python
 
    X = pd.Series(np.random.choice(['foo', 'bar'], size=(10,)))
-   X = X.astype('category', categories=['foo', 'bar'])
```
import first

jreback commented on doc/source/merging.rst:
```diff
@@ -842,8 +842,11 @@ The right frame.
 
 .. ipython:: python
 
```
same

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -22,6 +22,8 @@ Check the :ref:`API Changes <whatsnew_0210.api_breaking>` and :ref:`deprecations
 New features
 ~~~~~~~~~~~~
 
+- New user-facing :class:`pandas.api.types.CategoricalDtype` for specifying
```
move to the highlites (an provide the link to the sub-section); put the issue numbers there

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -106,6 +108,30 @@ This does not permit that column to be accessed as an attribute:
+expanded to include the ``categories`` and ``ordered`` attributes. A
+``CategoricalDtype`` can be used to specify the set of categories and
+orderedness of an array, independent of the data themselves. This can be useful,
+e.g., when converting string data to a ``Categorical``:
```
put issue refs here

jreback replied:
this

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -106,6 +108,30 @@ This does not permit that column to be accessed as an attribute:
+
+See :ref:`CategoricalDtype <categorical.categoricaldtype>` for more.
+
 .. _whatsnew_0210.enhancements.other:
```
looks like tslib.html was included somehow?

jreback commented on pandas/core/categorical.py:
```diff
@@ -474,13 +490,16 @@ def from_codes(cls, codes, categories, ordered=False):
                 "codes need to be convertible to an arrays of integers")
 
-        categories = cls._validate_categories(categories)
+        # have to use the instance, not property
```
shouldn't this me a method on the Dtype itself? (which does't the coercion via Index)?

TomAugspurger replied:
> shouldn't this me a method on the Dtype itself? 

You mean the codes -> categories? Something like

```python
class CategoricalDtype:
    def _categories_from_codes(self, codes: Array[Int]) -> Array[Any]:
        ...
```

I'll try that out. The current way is a bit odd.

jreback replied:
I mean having to wrap an ``Index`` around it *here*; the ``_validate_categories`` should do that (IOW be able to accept a list-like), (and it should return an Index)

jreback replied:
maybe call

``CategoricalDtype._validate_categories(categories)`` I think its more clear

jreback reviewed (changes requested):
some fb, more later

jreback commented on doc/source/categorical.rst:
```diff
@@ -96,12 +96,20 @@ By passing a :class:`pandas.Categorical` object to a `Series` or assigning it to
-    s_cat = s.astype("category", categories=["b","c","d"], ordered=False)
+    s = pd.Series(["a", "b", "c", "a"])
+    cat_type = pd.api.types.CategoricalDtype(categories=["b", "c", "d"],
+                                             ordered=False)
```
just do the import above

jreback commented on doc/source/categorical.rst:
```diff
@@ -140,6 +148,62 @@ constructor to save the factorize step during normal constructor mode:
+:class:`pandas.Categorical` is created.
+
+.. ipython:: python
+
```
show the import

jreback commented on doc/source/categorical.rst:
```diff
@@ -140,6 +148,62 @@ constructor to save the factorize step during normal constructor mode:
+
+A :class:`~pandas.api.types.CategoricalDtype` can be used in any place pandas
+expects a `dtype`. For example :func:`pandas.read_csv`,
+:func:`pandas.DataFrame.astype`, or the Series constructor.
```
or in the Series constructor

jreback commented on doc/source/categorical.rst:
```diff
@@ -140,6 +148,62 @@ constructor to save the factorize step during normal constructor mode:
+expects a `dtype`. For example :func:`pandas.read_csv`,
+:func:`pandas.DataFrame.astype`, or the Series constructor.
+
+As a convenience, you can use the string `'category'` in place of a
```
needs double back-ticks around ``'category'``

jreback commented on doc/source/categorical.rst:
```diff
@@ -140,6 +148,62 @@ constructor to save the factorize step during normal constructor mode:
+As a convenience, you can use the string `'category'` in place of a
+:class:`~pandas.api.types.CategoricalDtype` when you want the default behavior of
+the categories being unordered, and equal to the set values present in the
+array. On other words, ``dtype='category'`` is equivalent to
```
In other words

jreback commented on pandas/core/categorical.py:
```diff
@@ -250,20 +250,15 @@ class Categorical(PandasObject):
         if is_categorical_dtype(values):
 
-            # we are either a Series or a CategoricalIndex
+            # we are either a Series, CategoricalIndex or CategoricalDtype
```
this can't be true, values must be a value type and not a dtype (could be a Categorical though I think)

jreback commented on pandas/core/categorical.py:
```diff
@@ -325,12 +320,13 @@ def __init__(self, values, categories=None, ordered=False, fastpath=False):
+            dtype = CategoricalDtype(categories, ordered)
+            codes = _get_codes_for_values(values, dtype.categories)
 
             # TODO: check for old style usage. These warnings should be removes
```
maybe in followup (do we have an issue for this?). we should remove this and instead raise I think.

TomAugspurger replied:
Didn't see one, opened https://github.com/pandas-dev/pandas/issues/17485

jreback commented on pandas/core/categorical.py:
```diff
@@ -596,21 +552,13 @@ def _set_categories(self, categories, fastpath=False):
-        categories = self._validate_categories(categories, fastpath=fastpath)
-        if (not fastpath and self._categories is not None and
-                len(categories) != len(self._categories)):
+        new = CategoricalDtype(categories, self.ordered, fastpath)
```
new_dtype

jreback commented on pandas/core/categorical.py:
```diff
@@ -667,9 +613,9 @@ def set_ordered(self, value, inplace=False):
         """
         inplace = validate_bool_kwarg(inplace, 'inplace')
-        self._validate_ordered(value)
+        new = CategoricalDtype(self.categories, ordered=value)
```
same

jreback commented on pandas/core/categorical.py:
```diff
@@ -1085,7 +1030,7 @@ def __setstate__(self, state):
 
         # Provide compatibility with pre-0.15.0 Categoricals.
         if '_categories' not in state and '_levels' in state:
-            state['_categories'] = self._validate_categories(state.pop(
```
side issue, I think we can remove this (do we have an issue for it)?

jreback replied:
maybe we should create an issue on cleaning up old pickle compat (IOW we are at 0.13 now I think); we could eaily move to 0.17

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -692,6 +692,21 @@ def is_dtype_equal(source, target):
 
+def _is_dtype_union_equal(source, target):
+    """
+    Check whether two arrays have compatible dtypes to do a unoin.
```
unoin -> union

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -692,6 +692,21 @@ def is_dtype_equal(source, target):
+    """
+    Check whether two arrays have compatible dtypes to do a unoin.
+    numpy types are checked with ``is_dtype_equal``. Extension types are
+    checked separately.
```
can you add Parameters section

this is a private module so could do: ``is_dtype_union_equal``

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -692,6 +692,21 @@ def is_dtype_equal(source, target):
+    if is_categorical_dtype(source) and is_categorical_dtype(target):
+        # ordered False for both
+        return source.ordered is target.ordered
+    else:
```
no else here

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -1671,7 +1686,9 @@ def _coerce_to_dtype(dtype):
     """
 
     if is_categorical_dtype(dtype):
-        dtype = CategoricalDtype()
```
can you add a doc-string

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -110,37 +110,144 @@ class CategoricalDtypeType(type):
 
-    def __new__(cls):
+    def __new__(cls, categories=None, ordered=False, fastpath=False):
+        from pandas.core.indexes.base import Index
```
you could do the import in ``if categories is not None`` section

TomAugspurger replied:
Is that a performance or style thing? Style-wise I prefer in-method imports to be at the top of the method.

jreback replied:
actually a perf thing. This is *always* imported. Not sure its an issue though.

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -110,37 +110,144 @@ class CategoricalDtypeType(type):
+        # space for when we have unknown categories to avoid a conflict
+        if self.categories is None:
+            if self.ordered:
+                return -1
```
neat trick

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -110,37 +110,144 @@ class CategoricalDtypeType(type):
+            data = self.categories._format_data(name=self.__class__.__name__)
+        return tpl.format(data, self.ordered)
+
+    def __repr__(self):
```
``__repr__`` = ``_str__``

jreback replied:
?

TomAugspurger replied:
The base `ExtensionDtyep` class handles all this actually. I've removed the `__repr__` definition since it was redundant.

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -110,37 +110,144 @@ class CategoricalDtypeType(type):
+            hash_array, _combine_hash_arrays, hash_tuples
+        )
+
+        categories = np.asarray(categories)
```
aren't we guaranteed to be an Index here?

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -110,37 +110,144 @@ class CategoricalDtypeType(type):
+            categories = list(categories)  # breaks if a np.array of categories
+            cat_array = hash_tuples(categories)
+        else:
+            if categories.dtype == 'O':
```
we should actually move this to hash_array I think (could be pre-cursor PR)

jreback replied:
?

TomAugspurger replied:
Looking into this now.

Any thoughts on what this does to the probability of hash collisions? If we use two distinct hashing algorithms, I'd expect probability of a collision is higher (though presumably still negligible?),

jreback replied:
neh hash collisions basically don't happen in the space of things we are doing (normally). however you can collide by construction, e.g. code doing something wrong..

In any event we should move this code (can be a followup, maybe make an issue for things to do)

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -110,37 +110,144 @@ class CategoricalDtypeType(type):
+            ])
+        else:
+            cat_array = [cat_array]
+        hashed = _combine_hash_arrays(iter(cat_array),
```
I would move this last piece of logic to a private function in hashing as well.

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -154,6 +261,65 @@ def construct_from_string(cls, string):
+        -------
+        categories : Index
+        """
+        from pandas.core.dtypes.generic import ABCIndexClass
```
this can actually be imported at the top (ABCIndexClass)

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -154,6 +261,65 @@ def construct_from_string(cls, string):
+        categories : Index
+        """
+        from pandas.core.dtypes.generic import ABCIndexClass
+        from pandas import Index
```
can move to 303

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -27,6 +27,7 @@
     is_integer,
     is_float,
     is_dtype_equal,
+    _is_dtype_union_equal,
```
see change above

jreback commented on pandas/core/internals.py:
```diff
@@ -139,14 +139,14 @@ def is_categorical_astype(self, dtype):
-            if dtype == CategoricalDtype():
-                return True
-
+        if dtype is Categorical or dtype is CategoricalDtype:
```
hmm, this seems to contract the

``elif is_categorical_dtype(dtype)``

IOW we DO allow ``CategoricalDtype``, just not ``Categorical/CategoricalIndex`` here

you can just use ``pandas_dtype`` I think

TomAugspurger replied:
Some test was failing without this earlier. I'll see if I can revert it and figure out what was going on.

TomAugspurger replied:
Ah, this is to disallow passing in the class `CategoricalDtype`, instead of an instance. Without the check, you get an object dtype when (incorrectly) passing `CategoricalDtype` (not an instance)

```python
In [17]: pd.Series([1, 2]).astype(CategoricalDtype)
Out[17]:
0    1
1    2
dtype: object
```

jreback commented on pandas/core/internals.py:
```diff
@@ -548,6 +548,18 @@ def _astype(self, dtype, copy=False, errors='raise', values=None,
+                                "`categories` or `ordered`. Use "
+                                "`dtype=CategoricalDtype(categories, ordered)`"
+                                " instead.")
+            kwargs = kwargs.copy()
```
I think let's deprecate this

TomAugspurger replied:
Will raise a followup issue.

jreback commented on pandas/tests/dtypes/test_dtypes.py:
```diff
@@ -524,3 +521,91 @@ def test_caching(self):
+    def test_equal_but_different(self):
+        c1 = CategoricalDtype([1, 2, 3])
+        c2 = CategoricalDtype([1., 2., 3.])
+        assert c1 is not c2
```
assert not equality as well

jreback commented on pandas/tests/dtypes/test_dtypes.py:
```diff
@@ -524,3 +521,91 @@ def test_caching(self):
+    ])
+    def test_categorical_equality(self, ordered, other, expected):
+        c1 = CategoricalDtype(['a', 'b'], ordered)
+        result = c1 == other
```
add test for raises on invalid type for ``ordered`` (e.g. 'foo')

jreback commented on pandas/tests/test_categorical.py:
```diff
@@ -1619,6 +1644,13 @@ def test_validate_inplace(self):
             with pytest.raises(ValueError):
                 cat.sort_values(inplace=value)
 
+    @pytest.mark.xfail
```
add a reason here

jreback reviewed (changes requested):
looks really good. some style / code comments. how is perf?

TomAugspurger commented:
> how is perf?

Currently there's some issues with our internal fastpaths like `CategoricalIndex._shallow_copy` and `_simple_new`. I think I'm unnecessarily creating another instance of `CategoricalDtype` instead of passing the existing on on. Will fix that tomorrow.

jorisvandenbossche commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -110,37 +111,144 @@ class CategoricalDtypeType(type):
     _cache = {}
 
-    def __new__(cls):
+    def __new__(cls, categories=None, ordered=False, fastpath=False):
```
I know we have a `fastpath` arg in several places, but I don't really like it to have those in public functions (unless they are useful for external developers, but then it needs to be documented). Would it be possible to have a separate classmethod `from_..` that does the fastpath? (didn't look at the cases where this is used)

jorisvandenbossche commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -110,37 +111,144 @@ class CategoricalDtypeType(type):
+            return hash(self) == hash(other)
+
+    def __unicode__(self):
+        tpl = u'CategoricalDtype({}ordered={})'
```
I would be explicit in that the first element is categories ("categories=[..]"). It makes it longer, but clearer I think (otherwise could look like data)

jorisvandenbossche commented on pandas/core/indexes/base.py:
```diff
@@ -2170,7 +2173,10 @@ def union(self, other):
-        if not is_dtype_equal(self.dtype, other.dtype):
+        # TODO: is_dtype_union_equal is a hack around
+        # 1. buggy joins with duplicates
+        # 2. CategoricalIndex lacking setops
```
is this something you want to solve in this PR? Or is there an issue for this?

TomAugspurger replied:
I have a fix for CategoricalIndex set operations,  but it relies on proper support for set ops with duplicates. That ones is more involved. I've added references to the GH issues.

jorisvandenbossche commented on pandas/util/testing.py:
```diff
@@ -1244,7 +1244,12 @@ def assert_series_equal(left, right, check_dtype=True,
+        if (is_categorical_dtype(left) and is_categorical_dtype(right) and
+                not check_categorical):
+            # compat with pandas 0.21.0 CategoricalDtype
+            pass
```
why is this needed?

TomAugspurger replied:
This test: https://github.com/pandas-dev/pandas/pull/16015/files#diff-5066b062c892df33c29b37a8c1c75f75L1093 

I don't recall the details at the moment...

jorisvandenbossche replied:
Ah, it seems this is rather logical at a second look, as if `check_categorical=False`, the dtypes of categorical should not be checked exactly. So I would just remove the comment? Or better clarify it?

jorisvandenbossche commented on pandas/core/categorical.py:
```diff
@@ -1991,8 +1972,7 @@ def is_dtype_equal(self, other):
         try:
-            return (self.categories.equals(other.categories) and
-                    self.ordered == other.ordered)
+            return hash(self.dtype) == hash(other.dtype)
```
Can't this use the `__eq__` method of the dtype ?

TomAugspurger replied:
This would run into the strangeness of `CategoricalDtype(None) == CategoricalDtype(['a', 'b'], True)` being true, which we don't want here.

jorisvandenbossche replied:
Ah, that's a good point. But is there then a way for a user to do this, apart from this hash checking?

TomAugspurger replied:
>   But is there then a way for a user to do this, apart from this hash checking?

What does "this" mean? Checking if two `CategoricalDtypes` are the same?

TomAugspurger replied:
Yeah, in that case they would have to use the same "trick" of checking that the hashes are equal.

I expect comparing two instances `CategoricalDtype` to be relatively rare / for advanced users. In almost all of these cases, I expect users to be doing `a.dtype == b.dtype`, where `a` and `b` are `Categorical`, in which case they *won't* hit this gotcha, since a `Categorical().dtype` will never have unknown categories. It's just "free-standing" `CategoricalDtype`, not attached to a `Categorical`, that can have unknown categories.

jorisvandenbossche replied:
Indeed. And as an advanced user you can always make sure to avoid having dtype objects without categories

jorisvandenbossche reviewed (commented):
Added some comments!

General remark, I find the fact that a CategoricalDtype instance can exist without any categories/ordered specified a bit strange. But that is needed to support the case of `dtype='category'` where those are inferred from the data?

TomAugspurger commented:
Thanks Joris, I'll try to address those all this afternoon.

> General remark, I find the fact that a CategoricalDtype instance can exist without any categories/ordered specified a bit strange. But that is needed to support the case of dtype='category' where those are inferred from the data?

To be clear, the `ordered` will default to `False`, just like before. Agreed that the `categories` potentially being none is strange. I don't think users should run into it that often though... Anytime you have an actual array of categorical data, the `categories` will not be None. The only case I can imagine is where you know ahead of time that your data is ordered and sorted, but you don't know the categories?

```python
pd.read_csv("foo.csv", dtype={"A": CategoricalDtype(None, ordered=True)})
```

Which seems like a recipe for future bugs to me.

TomAugspurger commented on pandas/core/categorical.py:
```diff
@@ -248,22 +249,30 @@ class Categorical(PandasObject):
+                 fastpath=False):
 
-        self._validate_ordered(ordered)
+        if dtype is not None:
```
I just added it in response to the some performance loss. For large categoricals with many distinct values, creating another `CategoricalDtype` (with uniqueness and nan checking) can be somewhat costly. In some cases we may already have a `dtype` that can be reused.

jreback replied:
I think this is fine, ``Categorical`` is much like a Series/array-like which accepts ``dtype``

jreback replied:
maybe we can remove ``fastpath`` entirely now that we accept ``dtype`` (so maybe deprecate it)?

TomAugspurger replied:
Yeah, I'll do that as a followup.

TomAugspurger commented:
@jorisvandenbossche Agreed about not having a `fastpath` in the public API. see https://github.com/pandas-dev/pandas/pull/16015/commits/91752fe87f6cd5c06b7ee95b7fabb7534631a0df

Initially we used `__new__` for initialization, since we cached the instance (even though it didn't take parameters). That commit cleans things up a bit.

jreback commented on doc/source/api.rst:
```diff
@@ -646,7 +646,10 @@ strings and apply several methods to it. These can be accessed like
 
-If the Series is of dtype ``category``, ``Series.cat`` can be used to change the the categorical
+.. autoclass:: api.types.CategoricalDtype
+   :members: categories, ordered
```
yep, we should use the auto* things more readily in other places. Maybe make an issue about this.

jorisvandenbossche replied:
I don't think we in general need to do this, as for most functions/methods, we already have the generated pages to link to

TomAugspurger replied:
I just did this instead of `autosummary` since there are a bunch of unrelated methods that are just there for NumPy duck-typing.

jreback commented on doc/source/categorical.rst:
```diff
@@ -133,6 +143,70 @@ constructor to save the factorize step during normal constructor mode:
+
+A categorical's type is fully described by
+
+1. its categories: a sequence of unique values and no missing values
```
maybe show as

```
1. ``categories``: ....
2. ``ordered``: .....
```

as these are the actual kwargs

jreback commented on doc/source/categorical.rst:
```diff
@@ -133,6 +143,70 @@ constructor to save the factorize step during normal constructor mode:
+expects a `dtype`. For example :func:`pandas.read_csv`,
+:func:`pandas.DataFrame.astype`, or in the Series constructor.
+
+As a convenience, you can use the string ``'category'`` in place of a
```
maybe put this in a warning/note as this is the backward compat aspect here. (and then refer to this in the whatsnew)

jorisvandenbossche replied:
what is back incompat about doing `dtype='category'` ?

TomAugspurger replied:
I think Jeff was saying that was the only way to do things in the past (backwards compatible). But I don't want to give the impression that we're thinking about deprecating `dtype='category'`.

jorisvandenbossche replied:
Yes, but you didn't do that actual deprecation (or not yet)?

jorisvandenbossche replied:
> > Yes, but you didn't do that actual deprecation (or not yet)?

> I don't think this is desirable, 'category' is a nice convenience

Sorry, I was speaking about `astype(dtype='category', categories=[...])`. I suppose I misread Jeff's comment (as it was about backwards compatibility, I thought it was about that part, as this is the only part that might be back incompat, *if* we deprecate `categories=..` (not `dtype='category'`)

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -10,6 +10,8 @@ users upgrade to this version.
 
 - Integration with `Apache Parquet <https://parquet.apache.org/>`__, including a new top-level :func:`read_parquet` and :func:`DataFrame.to_parquet` method, see :ref:`here <io.parquet>`.
+- New user-facing :class:`pandas.api.types.CategoricalDtype` for specifying
+  categoricals independent of the data (:issue:`14711`, :issue:`15078`)
```
I would add the link the sub-section (and then list the issues there instead)

jreback commented on pandas/core/categorical.py:
```diff
@@ -202,6 +202,7 @@ class Categorical(PandasObject):
         categorical, read only.
     ordered : boolean
         Whether or not this Categorical is ordered.
+    dtype : CategoricalDtype
```
versionadded

jreback commented on pandas/core/categorical.py:
```diff
@@ -248,22 +249,30 @@ class Categorical(PandasObject):
-            # fast path
+            if dtype is None:
+                dtype = CategoricalDtype(categories, ordered)
             self._codes = coerce_indexer_dtype(values, categories)
```
IOW, you can move this *above* the dtype since fastpath just generates this

TomAugspurger replied:
Moving this up made the code further down messier, since the `categories` passed to `CategoricalDtype` need to be inferred from the `values`.

jreback commented on pandas/core/categorical.py:
```diff
@@ -313,7 +322,8 @@ def __init__(self, values, categories=None, ordered=False, fastpath=False):
 
-            categories = self._validate_categories(categories)
+            if dtype is None or isinstance(dtype, str):
+                dtype = CategoricalDtype(categories, ordered)
```
make this ``CategoricaDtype.construct_from_string(dtype, categories=categories, ordered=ordered)`` I think (just for validation in case someone passes

```
Categorical(....., dtype='foo')
```

jorisvandenbossche replied:
I think that if a user passes `dtype='foo'`, this actually already raises before (as in the beginning of init the .categories and .ordered attributes of the dtype are accessed if it is not None). 
Might be good to actually have a test case for this.

jorisvandenbossche replied:
did you mean Categorical instead of CategoricalDtype ?

jreback replied:
I meant exactly what I wrote.

the handling of ``dtype=`` in ``Categorical`` is duplicative and just needs to pass directly to ``CategoricalDtype`` (and that's why we need a constructor ``.from_string``)

jreback replied:
``CategoricalDtype([1, 2], dtype='categories')`` this should raise

CategoricalDtype([1, 2], dtype='category') might be ok though

TomAugspurger replied:
Whoops, I don't know what I was thinking when I wrote https://github.com/pandas-dev/pandas/pull/16015/files#r139390429

Here's the behavior (only `dtype=CategoricalDtype(...)` or `dtype='category'` works):

```python
In [2]: pd.Categorical([1, 2])
Out[2]:
[1, 2]
Categories (2, int64): [1, 2]

In [3]: pd.Categorical([1, 2], dtype='category')
Out[3]:
[1, 2]
Categories (2, int64): [1, 2]

In [4]: pd.Categorical([1, 2], dtype='categories')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-480c8cb555e4> in <module>()
----> 1 pd.Categorical([1, 2], dtype='categories')

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/categorical.py in __init__(self, values, categories, ordered, dtype, fastpath)
    240                     dtype = CategoricalDtype(categories, ordered)
    241                 else:
--> 242                     raise ValueError("Unknown `dtype` {}".format(dtype))
    243             elif categories is not None or ordered is not None:
    244                 raise ValueError("Cannot specify both `dtype` and `categories`"

ValueError: Unknown `dtype` categories
``

What's the argument for `CategoricalDtype(..., dtype='category')`? At the moment, I haven't added a dtype to `CategoricalDtype.__init__`. I'm not sure whether or not to do it, and if we do accept `CategoricalDtype` in `CategoricalDtype.__init__` then should it be the first argument `categories`, or a separate keyword? I'm going through the PR again, and I'll note all the places I have to do `isinstance(foo, CategoricalDtype)` before passing it to the constructor.

I'm also not sure changing `CategoricalDtype.construct_from_string` to accept categories (is that what you were suggisting @jreback?). Currently, it's just

```python
In [12]: pd.api.types.CategoricalDtype.construct_from_string('category')
Out[12]: CategoricalDtype(categories=None, ordered=False)
```

Everything else raises.

jreback commented on pandas/core/categorical.py:
```diff
@@ -608,21 +578,17 @@ def _set_categories(self, categories, fastpath=False):
-        categories = self._validate_categories(categories, fastpath=fastpath)
-        if (not fastpath and self._categories is not None and
-                len(categories) != len(self._categories)):
+        if fastpath:
```
lets deprecate ``fastpath`` (and convert fastpath -> dtype construction, and instead allow ``dtype`` to be passed

jreback commented on pandas/core/indexes/category.py:
```diff
@@ -161,16 +173,27 @@ def _simple_new(cls, values, name=None, categories=None, ordered=None,
+        # we want to reuse self.dtype if possible, i.e. neither are
+        # overridden.
+        if dtype is not None and (categories is not None or
+                                  ordered is not None):
```
msg?

jreback commented on pandas/tests/dtypes/test_common.py:
```diff
@@ -545,10 +545,11 @@ def test_is_complex_dtype():
+    (pd.Categorical(['a', 'b']), CategoricalDtype(['a', 'b'])),
+    (pd.CategoricalIndex(['a', 'b']).dtype, CategoricalDtype(['a', 'b'])),
+    (pd.CategoricalIndex(['a', 'b']), CategoricalDtype(['a', 'b'])),
+    (CategoricalDtype(), CategoricalDtype()),
```
leave original ``pd.Categorical(['a', 'b']).dtype, CategoricalDtype())`` which should still work

jreback commented on pandas/tests/dtypes/test_dtypes.py:
```diff
@@ -524,3 +521,96 @@ def test_caching(self):
+        assert result == expected
+
+    def test_invalid_raises(self):
+        with tm.assert_raises_regex(TypeError, 'ordered'):
```
might add a test for passing just a string for the first argument (should fail)
e.g.
``CategoricalDtype('category')`` (see my comment about a ``from_string`` above, which we should add for compat.

jreback reviewed (changes requested):
look good. a few changes.

TomAugspurger commented:
Correct, and I don't plan to deprecate `dtype'category'`.

On Mon, Sep 18, 2017 at 5:55 AM, Joris Van den Bossche <
notifications@github.com> wrote:

> *@jorisvandenbossche* commented on this pull request.
> ------------------------------
>
> In doc/source/categorical.rst
> <https://github.com/pandas-dev/pandas/pull/16015#discussion_r139390284>:
>
> > +should be inferred from whatever is present in the data when the
> +:class:`pandas.Categorical` is created.
> +
> +.. ipython:: python
> +
> +   from pandas.api.types import CategoricalDtype
> +
> +   CategoricalDtype(['a', 'b', 'c'])
> +   CategoricalDtype(['a', 'b', 'c'], ordered=True)
> +   CategoricalDtype()
> +
> +A :class:`~pandas.api.types.CategoricalDtype` can be used in any place pandas
> +expects a `dtype`. For example :func:`pandas.read_csv`,
> +:func:`pandas.DataFrame.astype`, or in the Series constructor.
> +
> +As a convenience, you can use the string ``'category'`` in place of a
>
> Yes, but you didn't do that actual deprecation (or not yet)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/16015#discussion_r139390284>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHItygdUADe7vz66Heen2kIDzUzdl9ks5sjkwVgaJpZM4M-hoW>
> .
>

jreback commented:
> Yes, but you didn't do that actual deprecation (or not yet)?

I don't think this is desirable, 'category' is a nice convenience

TomAugspurger committed:
update

TomAugspurger committed:
Consistent CategoricalDtype use in Categorical init

Get a valid instance of `CategoricalDtype` as early as possible, and use that
throughout.

TomAugspurger commented:
https://github.com/pandas-dev/pandas/pull/16015/commits/416d1d7b9ce81287a1b7f2802751dca1fdeb5084 has the changes to `Categorical.__init__` that you requested @jreback. It's a bit messy since there are basically 3 ways for specifying the categories

1. passing `dtype=CategoricalDtype(...)`
2. passing `categories=...`
3. passing a `Categorical` for `values`

I think all three are useful and (hopefully) well-tested now. I'll update the docstring to indicate the priority (basically moving that comment to the docstring)

TomAugspurger committed:
PEP8 fixes

jreback commented on doc/source/categorical.rst:
```diff
@@ -133,6 +143,73 @@ constructor to save the factorize step during normal constructor mode:
+
+   Since ``dtype='category'`` is essentially ``CategoricalDtype(None, False)``,
+   and since all instances ``CategoricalDtype`` compare equal to ``'`category'``,
+   all instances of ``CategoricalDtype`` compare equal to a ``CategoricalDtype(None)``
```
so you should say ``CategoricalDtype(None, False)`` at the end.

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -110,37 +111,161 @@ class CategoricalDtypeType(type):
+            return hash(self) == hash(other)
+
+    def __unicode__(self):
+        tpl = u'CategoricalDtype(categories={}ordered={})'
```
we have this hacky thing in Categorical to repr only 10 categories. CategoricalIndex actually correctly uses the option ``max_categories``, so should do that here (again can do as a followup)

jreback commented on pandas/tests/test_categorical.py:
```diff
@@ -142,6 +142,26 @@ def test_constructor_empty(self):
+        tm.assert_index_equal(result.categories, expected)
+        assert result.ordered is False
+
+    def test_constructor_tuples_datetimes(self):
```
this is janky. I don't think we should actually allow it. Where did this example come from? So we allow tuples as categories, I guess so.

TomAugspurger replied:
This was extracted from a groupby test that was failing when I improperly handled tuples of level-1. Agreed that it's weird, but I'm inclined to support it.

jreback reviewed (approved):
just a couple of comments, which can be put in a followup. let's merge on green.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -89,6 +91,30 @@ This does not raise any obvious exceptions, but also does not create a new colum
+expanded to include the ``categories`` and ``ordered`` attributes. A
+``CategoricalDtype`` can be used to specify the set of categories and
+orderedness of an array, independent of the data themselves. This can be useful,
+e.g., when converting string data to a ``Categorical`` (:issue:`14711`, :issue:`15078`):
```
there is 1 more issue you listed in the top of the PR

TomAugspurger committed:
Final doc fixups

TomAugspurger commented:
Thanks. I fixed up the doc comments. I'll merge on green and do the other
followups later today.

On Fri, Sep 22, 2017 at 7:31 PM, Jeff Reback <notifications@github.com>
wrote:

> *@jreback* commented on this pull request.
> ------------------------------
>
> In doc/source/whatsnew/v0.21.0.txt
> <https://github.com/pandas-dev/pandas/pull/16015#discussion_r140620020>:
>
> > @@ -89,6 +91,30 @@ This does not raise any obvious exceptions, but also does not create a new colum
>
>  Setting a list-like data structure into a new attribute now raise a ``UserWarning`` about the potential for unexpected behavior. See :ref:`Attribute Access <indexing.attribute_access>`.
>
> +.. _whatsnew_0210.enhancements.categorical_dtype:
> +
> +``CategoricalDtype`` for specifying categoricals
> +^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> +
> +:class:`pandas.api.types.CategoricalDtype` has been added to the public API and
> +expanded to include the ``categories`` and ``ordered`` attributes. A
> +``CategoricalDtype`` can be used to specify the set of categories and
> +orderedness of an array, independent of the data themselves. This can be useful,
> +e.g., when converting string data to a ``Categorical`` (:issue:`14711`, :issue:`15078`):
>
> there is 1 more issue you listed in the top of the PR
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/16015#pullrequestreview-64725698>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIlwjFqIXM4vMTgsVZ20tYOzp9CQSks5slFFogaJpZM4M-hoW>
> .
>

TomAugspurger committed:
Fixup set_categories inplace test

TomAugspurger committed:
PEP8

jreback commented:
nice patch @TomAugspurger !
